### ISSUE TITLE
Combination of Caching and Paging does not work correctly

### ISSUE THREAD

**Issue by salascz** [pnp/pnpjs#2836]
### Major Version

3.x

### Minor Version Number

3.10.0

### Target environment

All

### Additional environment details

Node - 16
Package manager - yarn

### Expected or Desired Behavior

I am trying to combine paging and caching behaviour in my application. Looks like I am facing the same issue as described [here](https://github.com/pnp/pnpjs/issues/2793)

I have a list with approx. 1000 items. I am trying to get items paged by 500:

`.top(500).getPaged<TItem[]>();`

using session storage:

```
.using(Caching(
   {
    store: 'session',
    expireFunc: () => dateAdd(new Date(), 'hour', 1) ?? new Date(),
  }
));
```

### Observed Behavior

The initial request works correctly:
 - it executes two calls
 - first call contains `odata.nextLink` attribute
 - data are stored to session cache (two records, one for each call). 
    ![image](https://github.com/pnp/pnpjs/assets/5662939/f0788deb-ebc2-4d19-a1b2-b9934974bb43)
    ![image](https://github.com/pnp/pnpjs/assets/5662939/bc3d64cd-87eb-42db-887a-571f258bd3c2)
    Observation: `odata.nextLink` is changed to `nextUrl`

Repeating requests (e.g. page reload when the cache is not outdated) does not work correctly:
- the `getPaged` method returns first result from the cache
- the `hasNext` attribute is `undefined`  
   ![image](https://github.com/pnp/pnpjs/assets/5662939/39e57f64-67a0-4408-bb59-597c874ecfba)
- but there is a `nextUrl` attribute 
   ![image](https://github.com/pnp/pnpjs/assets/5662939/9630f893-6fc0-4fd4-b55c-26b5281c4b45)

Am I missing something? It almost look like there is some transformation missing for results from cache...

### Steps to Reproduce

my helpers methods

```
  public static async getItemsPaged<TItem>(query: IItems, pageSize: number): Promise<TItem[]> {
    const results: TItem[] = [];
    let dataPaged: PagedItemCollection<TItem[]> | null = await QueryHelper.getItemsPagedFirstPage<TItem>(query, pageSize);
    results.push(...dataPaged.results);
    while (dataPaged?.hasNext) { // undefined when loaded from cache
      dataPaged = await dataPaged.getNext();
      if (!dataPaged) {
        break;
      }
      results.push(...dataPaged.results);
    }
    return results;
  }


  public static async getItemsPagedFirstPage<TItem>(query: IItems, pageCount: number): Promise<PagedItemCollection<TItem[]>> {
    // add order by ID for successful paging
    return query
      .orderBy('ID')
      .top(pageCount).getPaged<TItem[]>();
  }
```

**Comment by juliemturner** [comment]
Yes, Basic caching is meant to work for individual calls and is a generic helper method. I am not 100% sure how to do this with a behavior, but I suspect adding a custom cache key would solve it, you'd have to test it: https://pnp.github.io/pnpjs/queryable/behaviors/#cachekey

If that's not what you're looking for then you would need to implement a custom cache function and or do it in your own code.

**Comment by salascz** [comment]
Hi @juliemturner,
thanks for quick reply. 

Not sure if the **custom cache key** helps in this case.  Did you mean to set different cache key before each getNext() call? (and test if such a key exists by custom code)
BTW: setting the custom cache key prior to first call is even worse (same key is used for initial and `getNext()` calls)

I understand such a solution is not supported. Would it be possible to mention this as an unsupported scenario for `getPaged()` in documentation?

Looks like caching on a different level would be easier (after I get all data initially).

Thanks

**Comment by juliemturner** [comment]
I would agree, I would cache in my own code. Yes, we can mark things as not supported by caching.

**Comment by bcameron1231** [comment]
Yea this one is a bit tricky because we store the results of response, but aren't re-casting it back out from cache, so the getter for HasNext isn't initialized. We can take a look at possibly re-hydrating the objects from cache.

**Comment by bcameron1231** [comment]
This has been deployed in 3.21 release.
